test/test_ensemble_builder/test_ensemble_builder.py

```diff
@@ -0,0 +1,554 @@
+) -> None:
+    """
+    Expects
+    -------
+    * Should have nothing in common between candidates and discarded
+    * Should filter out runs with no "ensemble" predictions
+    """
+    bad_runs = [make_run(predictions=None) for _ in range(5)]
+    dummy = make_run(dummy=True, loss=2)
+    good_run = make_run(predictions="ensemble", loss=1)
+
+    runs = bad_runs + [good_run]
+
+    candidates, discarded = builder.candidate_selection(runs, dummy)
+
+    assert len(set(candidates) & discarded) == 0
+    assert len(candidates) == 1
+    assert len(discarded) == len(bad_runs)
+    assert candidates[0].pred_path("ensemble").exists()
+
+
+def test_candidates_filters_runs_with_no_loss(
+    builder: EnsembleBuilder, make_run: Callable[..., Run]
+) -> None:
+    """
+    Expects
+    -------
+    * Should filter out runs with no loss
+    """
+    bad_runs = [make_run(loss=None) for _ in range(5)]
+    dummy_run = make_run(dummy=True, loss=2)
+    good_run = make_run(loss=1)
+
+    runs = bad_runs + [good_run]
+
+    candidates, discarded = builder.candidate_selection(runs, dummy_run)
+
+    assert len(candidates) == 1
+    assert len(discarded) == len(bad_runs)
+    assert candidates[0].loss == 1
+
+
+def test_candidates_filters_out_better_than_dummy(
+    builder: EnsembleBuilder,
+    make_run: Callable[..., Run],
+) -> None:
+    """
+    Expects
+    -------
+    * Should filter out runs worse than dummy
+    """
+    bad_runs = [make_run(loss=1) for _ in range(2)]
+    dummy_run = make_run(dummy=True, loss=0)
+    good_runs = [make_run(loss=-1) for _ in range(3)]
+
+    runs = bad_runs + good_runs
+
+    candidates, discarded = builder.candidate_selection(
+        runs, dummy_run, better_than_dummy=True
+    )
+
+    assert set(candidates)
+
+    assert len(candidates) == 3
+    assert all(run.loss < dummy_run.loss for run in candidates)
+
+    assert len(discarded) == 2
+    assert all(run.loss >= dummy_run.loss for run in discarded)
+
+
+def test_candidates_uses_dummy_if_no_candidates_better(
+    builder: EnsembleBuilder,
+    make_run: Callable[..., Run],
+) -> None:
+    """
+    Expects
+    -------
+    * If no run is better than a dummy run, the candidates will then consist
+      of the dummy runs.
+    """
+    runs = [make_run(loss=10) for _ in range(10)]
+    dummies = [make_run(dummy=True, loss=0) for _ in range(2)]
+
+    candidates, discarded = builder.candidate_selection(
+        runs,
+        dummies,
+        better_than_dummy=True,
+    )
+
+    assert len(candidates) == 2
+    assert all(run.is_dummy() for run in candidates)
+
+
+@parametrize("nbest", [0, 1, 5, 1000])
+def test_candidates_nbest_int(
+    builder: EnsembleBuilder,
+    make_run: Callable[..., Run],
+    nbest: int,
+) -> None:
+    """
+    Expects
+    -------
+    * Should only select the nbest candidates
+    * They should be ordered by loss
+    """
+    n = 10
+    expected = int(bound(nbest, bounds=(1, n)))
```

Comment:
Should there also be a max(1, ...) around this to ensure we expect at least one model?